components/compositing/refresh_driver.rs

```diff
@@ -2,129 +2,141 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
 
-use std::cell::Cell;
-use std::collections::hash_map::Values;
+use std::cell::{Cell, RefCell};
+use std::rc::Rc;
 use std::sync::Arc;
 use std::sync::atomic::{AtomicBool, Ordering};
 use std::thread::{self, JoinHandle};
 use std::time::Duration;
 
-use base::id::WebViewId;
 use constellation_traits::EmbedderToConstellationMessage;
 use crossbeam_channel::{Sender, select};
-use embedder_traits::EventLoopWaker;
+use embedder_traits::{EventLoopWaker, RefreshDriver};
 use log::warn;
 use timers::{BoxedTimerCallback, TimerEventRequest, TimerScheduler};
 
+use crate::IOCompositor;
 use crate::compositor::RepaintReason;
 use crate::webview_renderer::WebViewRenderer;
 
-const FRAME_DURATION: Duration = Duration::from_millis(1000 / 120);
-
-/// The [`RefreshDriver`] is responsible for controlling updates to aall `WebView`s
-/// onscreen presentation. Currently, it only manages controlling animation update
-/// requests.
-///
-/// The implementation is very basic at the moment, only requesting new animation
-/// frames at a constant time after a repaint.
-pub(crate) struct RefreshDriver {
-    /// The channel on which messages can be sent to the Constellation.
-    pub(crate) constellation_sender: Sender<EmbedderToConstellationMessage>,
-
-    /// Whether or not we are currently animating via a timer.
-    pub(crate) animating: Cell<bool>,
-
-    /// Whether or not we are waiting for our frame timeout to trigger
-    pub(crate) waiting_for_frame_timeout: Arc<AtomicBool>,
-
-    /// A [`TimerThread`] which is used to schedule frame timeouts in the future.
-    timer_thread: TimerThread,
-
-    /// An [`EventLoopWaker`] to be used to wake up the embedder when it is
-    /// time to paint a frame.
+/// The [`BaseRefreshDriver`] is a "base class" for [`RefreshDriver`] trait
+/// implementations. It encapsulates shared behavior so that it does not have to be
+/// implemented by all trait implementations. It is responsible for providing
+/// [`RefreshDriver`] implementations with a callback that is used to wake up the event
+/// loop and trigger frame readiness.
+pub(crate) struct BaseRefreshDriver {
+    waiting_for_frame: Arc<AtomicBool>,
```

Comment:
Could we add doc comments for these fields?

Reply:
Sure. Done.